db/skipped_sequence.go

```diff
@@ -181,34 +259,94 @@ func (s *SkippedSequenceSlice) removeSeq(x uint64) error {
 		s.list = slices.Delete(s.list, index, index+1)
 		return nil
+	} else {
+		// range entry handling
+		// if x-1 == range start seq we need to replace elem at index-1 with single entry, then alter range start seq to x+1
+		rangeElem := s.list[index]
+		if x-1 == rangeElem.getStartSeq() {
+			if rangeElem.getNumSequencesInEntry() == 3 {
+				// add single entry at elem
+				newElem := NewSingleSkippedSequenceEntry(rangeElem.getStartSeq())
+				newElem.setTimestamp(rangeElem.getTimestamp())
+				s.list[index] = newElem
+				// add single entry at elem + 1
+				newElem = NewSingleSkippedSequenceEntry(rangeElem.getLastSeq())
+				newElem.setTimestamp(rangeElem.getTimestamp())
+				s.insert(index+1, newElem)
+			} else {
+				newElem := NewSingleSkippedSequenceEntry(x - 1)
+				newElem.setTimestamp(rangeElem.getTimestamp())
+				s.insert(index, newElem)
+				rangeElem.setStartSeq(x + 1)
+			}
+			return nil
+		}
+
+		if x+1 == rangeElem.getLastSeq() {
+			newElem := NewSingleSkippedSequenceEntry(x + 1)
+			newElem.setTimestamp(rangeElem.getTimestamp())
+			s.insert(index+1, newElem)
+			rangeElem.setLastSeq(x - 1)
+			return nil
+		}
+
+		// if x != start or end seq on the range we need to alter the range and index and insert a new range at index + 1
+		if rangeElem.getStartSeq() < x && rangeElem.getLastSeq() > x {
+			// split index range
+			newElem := NewSkippedSequenceRangeEntry(x+1, rangeElem.getLastSeq())
+			newElem.setTimestamp(rangeElem.getTimestamp())
+			s.insert(index+1, newElem)
+			rangeElem.setLastSeq(x - 1)
+			return nil
+		}
+		// x is equal to start or end seq on the element so simply alter range start or end values
+		if rangeElem.getStartSeq() == x {
+			rangeElem.setStartSeq(x + 1)
+		} else {
+			rangeElem.setLastSeq(x - 1)
+		}
 	}
-	// more range handling here CBG-3853, temporarily error as we shouldn't get to this point at this time
-	return fmt.Errorf("entered range handling code")
+	return nil
 }
 
 // insert will insert element in middle of slice maintaining order of rest of slice
 func (s *SkippedSequenceSlice) insert(index int, entry SkippedSequenceListEntry) {
-	s.lock.Lock()
-	defer s.lock.Unlock()
 	s.list = append(s.list, nil)
 	copy(s.list[index+1:], s.list[index:])
 	s.list[index] = entry
 }
 
 // PushSkippedSequenceEntry will append a new skipped sequence entry to the end of the slice, if adding a contiguous
 // sequence function will expand the last entry of the slice to reflect this
-func (s *SkippedSequenceSlice) PushSkippedSequenceEntry(entry *SingleSkippedSequence) {
+func (s *SkippedSequenceSlice) PushSkippedSequenceEntry(entry SkippedSequenceListEntry) {
 	s.lock.Lock()
 	defer s.lock.Unlock()
 
 	if len(s.list) == 0 {
 		s.list = append(s.list, entry)
 		return
 	}
-	// adding contiguous sequence handling here, pending CBG-3853
-	s.list = append(s.list, entry)
+	// get index of last entry + last seq of entry
+	index := len(s.list) - 1
+	lastEntryLastSeq := s.list[index].getLastSeq()
```

Comment:
Have you thought about what we want to do with timestamps when the incoming entry is combined with the existing?  There are impacts to SkippedSequenceSlice eviction in either case, but I think using the incoming entry's timestamp is probably more correct on the whole.  Would be interested in your thoughts.

Reply:
I agree upon reflection. I have added a new method to do this on the interface.